Packages/com.unity.inputsystem/InputSystem/InputManager.cs

```diff
@@ -3271,6 +3271,10 @@ internal void RestoreDevicesAfterDomainReload()
                         {
                             device = AddDevice(deviceState.description, throwIfNoLayoutFound: true,
                                 deviceId: deviceState.deviceId, deviceFlags: deviceState.flags);
+
+                            // Re-set the name to the stored name. Otherwise we lose the sorting of 
+                            // devices to remote players, which is encoded in the name.
+                            device.m_Name = new InternedString(deviceState.name);
```

Comment:
Wondering if we should put this inside `if ((deviceState.flags & InputDevice.DeviceFlags.Remote) != 0)`.

Overall, the logic here is more a workaround. Think the proper solution would be to store a local device ID to remote device ID mapping and get rid of the name dependency.

Reply:
What do you exactly mean by "get rid of the name dependency" and "proper solution would be to store a local device ID to remote device ID mapping and get rid of the name dependency".

-Use the stored mapping to restore the name correctly?
-Don't encode player id in the name at all, but change the input debugger code to sort devices in the UI using some other information?

Reply:
What do you think of adding a participant Id int member to the InputDevice class instead of encoding it in the name, like this:

```
diff --git a/Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs b/Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs
index 3fb4b3c2..8b807f40 100644
--- a/Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs
+++ b/Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs
@@ -39,6 +39,7 @@ namespace UnityEngine.Experimental.Input
     public class InputDevice : InputControl
     {
         public const int kInvalidDeviceId = 0;
+        public const int kLocalParticipantId = 0;
         internal const int kInvalidDeviceIndex = -1;
 
         /// <summary>
@@ -264,6 +265,7 @@ namespace UnityEngine.Experimental.Input
         public InputDevice()
         {
             m_Id = kInvalidDeviceId;
+            m_ParticipantId = kLocalParticipantId;
             m_DeviceIndex = kInvalidDeviceIndex;
         }
 
@@ -402,6 +404,7 @@ namespace UnityEngine.Experimental.Input
 
         internal DeviceFlags m_DeviceFlags;
         internal int m_Id;
+        internal int m_ParticipantId;
         internal int m_DeviceIndex; // Index in InputManager.m_Devices.
         internal InputDeviceDescription m_Description;
 
diff --git a/Packages/com.unity.inputsystem/InputSystem/Devices/Remote/InputRemoting.cs b/Packages/com.unity.inputsystem/InputSystem/Devices/Remote/InputRemoting.cs
index d699afd6..c0e6648b 100644
--- a/Packages/com.unity.inputsystem/InputSystem/Devices/Remote/InputRemoting.cs
+++ b/Packages/com.unity.inputsystem/InputSystem/Devices/Remote/InputRemoting.cs
@@ -516,8 +516,9 @@ namespace UnityEngine.Experimental.Input
                 InputDevice device;
                 try
                 {
-                    device = receiver.m_LocalManager.AddDevice(data.layout,
-                        string.Format("Remote{0}::{1}", msg.participantId, data.name));
+                    device = receiver.m_LocalManager.AddDevice(data.layout);
+                    device.m_ParticipantId = msg.participantId;
                 }
                 catch (Exception exception)
                 {
diff --git a/Packages/com.unity.inputsystem/InputSystem/Editor/InputDebuggerWindow.cs b/Packages/com.unity.inputsystem/InputSystem/Editor/InputDebuggerWindow.cs
index 13f417df..871c0183 100644
--- a/Packages/com.unity.inputsystem/InputSystem/Editor/InputDebuggerWindow.cs
+++ b/Packages/com.unity.inputsystem/InputSystem/Editor/InputDebuggerWindow.cs
@@ -341,7 +341,7 @@ namespace UnityEngine.Experimental.Input.Editor
                     foreach (var player in EditorConnection.instance.ConnectedPlayers)
                     {
                         var playerNode = AddChild(remoteDevicesNode, player.name, ref id);
-                        AddDevices(playerNode, devices, ref id, "Remote" + player.playerId + InputControlLayout.kNamespaceQualifier);
+                        AddDevices(playerNode, devices, ref id, player.playerId);
                     }
                 }
                 else
@@ -424,23 +424,18 @@ namespace UnityEngine.Experimental.Input.Editor
                 return root;
             }
 
-            private void AddDevices(TreeViewItem parent, IEnumerable<InputDevice> devices, ref int id, string namePrefix = null)
+            private void AddDevices(TreeViewItem parent, IEnumerable<InputDevice> devices, ref int id, int participantId = InputDevice.kLocalParticipantId)
             {
                 foreach (var device in devices)
                 {
-                    if (namePrefix != null)
-                    {
-                        if (!device.name.StartsWith(namePrefix))
-                            continue;
-                    }
-                    else if (device.name.Contains(InputControlLayout.kNamespaceQualifier))
-                        continue;
+                    if (device.m_ParticipantId != participantId)
+                        continue; 
 
                     var item = new DeviceItem
                     {
                         id = id++,
                         depth = parent.depth + 1,
-                        displayName = namePrefix != null ? device.name.Substring(namePrefix.Length) : device.name,
+                        displayName = device.name,
                         device = device,
                         icon = EditorInputControlLayoutCache.GetIconForLayout(device.layout),
                     };
diff --git a/Packages/com.unity.inputsystem/InputSystem/InputManager.cs b/Packages/com.unity.inputsystem/InputSystem/InputManager.cs
index 8e49e63e..cd3dd662 100644
--- a/Packages/com.unity.inputsystem/InputSystem/InputManager.cs
+++ b/Packages/com.unity.inputsystem/InputSystem/InputManager.cs
@@ -3120,6 +3120,7 @@ namespace UnityEngine.Experimental.Input
             public string variants;
             public string[] usages;
             public int deviceId;
+            public int participantId;
             public InputDevice.DeviceFlags flags;
             public InputDeviceDescription description;
 
@@ -3179,6 +3180,7 @@ namespace UnityEngine.Experimental.Input
                     layout = device.layout,
                     variants = device.variants,
                     deviceId = device.id,
+                    participantId = device.m_ParticipantId,
                     usages = usages,
                     description = device.m_Description,
                     flags = device.m_DeviceFlags
@@ -3271,10 +3273,7 @@ namespace UnityEngine.Experimental.Input
                         {
                             device = AddDevice(deviceState.description, throwIfNoLayoutFound: true,
                                 deviceId: deviceState.deviceId, deviceFlags: deviceState.flags);
-
-                            // Re-set the name to the stored name. Otherwise we lose the sorting of 
-                            // devices to remote players, which is encoded in the name.
-                            device.m_Name = new InternedString(deviceState.name);
+                            device.m_ParticipantId = deviceState.participantId;
                         }
                         else
                         {
```